### PR TITLE
chore(botocore): abstract integration from products in patch entrypoint

### PR DESCRIPTION
**Pull request by emmettbutler**
This pull request removes mentions of Spans and Tracing from the main botocore patching entrypoint, increasing the separation of concerns between the integration and Products. There are also some refactors aimed at deduplicating code to make it easier to abstract in this way.

## Checklist

- [x] Change(s) are motivated and described in the PR description
- [x] Testing strategy is described if automated tests are not included in the PR
- [x] Risks are described (performance impact, potential for breakage, maintainability)
- [x] Change is maintainable (easy to change, telemetry, documentation)
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed or label `changelog/no-changelog` is set
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/))
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))
- [x] If this PR changes the public interface, I've notified `@DataDog/apm-tees`.

## Reviewer Checklist

- [x] Title is accurate
- [x] All changes are related to the pull request's stated goal
- [x] Description motivates each change
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes
- [x] Testing strategy adequately addresses listed risks
- [x] Change is maintainable (easy to change, telemetry, documentation)
- [x] Release note makes sense to a user of the library
- [x] Author has acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

File: ddtrace/contrib/botocore/services/stepfunctions.py [hunk 1]
```diff
@@ -55,9 +55,6 @@ def inject_trace_to_stepfunction_input(params, span):
         if isinstance(input_obj, dict):
             input_obj["_datadog"] = {}
             HTTPPropagator.inject(span.context, input_obj["_datadog"])
-            input_json = json.dumps(input_obj)
```
Comment (brettlangdon) [inline ddtrace/contrib/botocore/services/stepfunctions.py:58]: why are we removing this?

**Review by brettlangdon — COMMENTED** [review]
I have a question about the stepfunctions change, otherwise lgtm

File: ddtrace/contrib/botocore/services/stepfunctions.py (hunk 1 above)
Comment (emmettbutler) [reply ddtrace/contrib/botocore/services/stepfunctions.py:58]: As far as I can tell it's not used, and modifying a mutable argument can lead to confusing side effects from the perspective of callers of this function. Is it used somewhere after all? @brettlangdon

Comment (brettlangdon) [reply ddtrace/contrib/botocore/services/stepfunctions.py:58]: this function modifies a mutable dict, so the caller expects this function to mutate the dict, not return a dict :/

Comment (emmettbutler) [reply ddtrace/contrib/botocore/services/stepfunctions.py]: I see what's happening now. I reverted the change. @brettlangdon

**Review by brettlangdon — APPROVED** [review]

**Pull request merged**